File: CS_Samples/dat_create.cs
///////////////////////////////////////////////////////////////////////////////
///
/// FuryUtils Sample
/// 
/// Compressing and archiving files into a Dat format file.
/// 
/// using:
/// 
///    Dat
///         class for reading/writing Dat archive files.
///    FuryException
///         exception class for reporting errors.
///
///
/// Suggested inputs from the testassets: pal8out.imm, pal8out.pam
/// 
///////////////////////////////////////////////////////////////////////////////

using carbon14.FuryStudio.Utils;

namespace carbon14.FuryStudio.CS_Samples
{
    internal class dat_create
    {
        public static void Execute (string fileName1, string fileName2, string archiveFileName)
        {
            Console.WriteLine("dat_create Sample");

            try
            {
                // Read a file to be archived into a buffer.
                byte[] fileBuffer1;
                using (FileStream fs = new FileStream(fileName1, FileMode.Open))
                {
                    fileBuffer1 = new byte[fs.Length];
                    fs.Read(fileBuffer1, 0, fileBuffer1.Length);
                }

                // Read a second file to be archived into a buffer.
                byte[] fileBuffer2;
                using (FileStream fs = new FileStream(fileName2, FileMode.Open))
                {
                    fileBuffer2 = new byte[fs.Length];
                    fs.Read(fileBuffer2, 0, fileBuffer2.Length);
                }

                byte[]? archiveBuffer;

                // Create a new Dat archive.
                using (Dat archive = new Dat())
                {
                    // Add the first file to the archive, with compression.
                    archive.Add("File1", fileBuffer1, true);
                    
                    // Add the second file to the archive without compression.
                    archive.Add("File2", fileBuffer2, false);

                    // Get a buffer with the archived data.
                    archiveBuffer = archive.Buffer;
                }

                // Save the archive buffer to a file.
                if (archiveBuffer != null)
                {
                    using (FileStream fs = new FileStream(archiveFileName, FileMode.Create))
                    {
                        fs.Write(archiveBuffer, 0, archiveBuffer.Length);
                    }
                }
            }
            catch (FuryException ex)
            {
                Console.WriteLine($"Error: {ex.ErrorCode} {ex.Message}");
            }
        }

    }
}

File: Utils/Cpp_Samples/dat_create.cpp
///////////////////////////////////////////////////////////////////////////////
// FuryUtils Sample
//
// Compressing and archiving files into a Dat format file
//
// using:
//
//    Dat
//        class for reading / writing Dat archive files.
//    Exceptions::Exception
//       exception class for reporting errors.
//
//
//   Suggested inputs from the testassets: pal8out.imm, pal8out.pam
//
///////////////////////////////////////////////////////////////////////////////

#include <ios>
#include <fstream>
#include "../include/FuryUtils.hpp"

void dat_create(const char * fileName1, const char * fileName2, const char * archiveFileName) {
	printf("dat_create Sample\n");
	
	try {

		// create a Dat archive
		FuryUtils::Archive::Dat df;

		{
			// Get a stream for the first file and load it into a buffer
			std::ifstream file(fileName1, std::ios::binary | std::ios::ate);
			std::streamsize size = file.tellg();
			file.seekg(0, std::ios::beg);
			std::vector<uint8_t> buffer((uint32_t)size);
			file.read((char *)(buffer.data()), size);

			// Add the buffer to the Dat archive with compression
			df.Add("File1", buffer, true);
			
			// Read the header from the Dat entry so that we can print 
			// some details about it.
			FuryUtils::Archive::DatHeader *dfh = df.Header(df.EntryCount() - 1);
			if (dfh->IsNotCompressed) {
				printf("%12s\tUncompressed\t%d\n", dfh->FileName, dfh->UncompressedSize);
			}
			else {
				printf("%12s\tCompressed - %d%%\t%d\t%d\n", dfh->FileName, (100 * dfh->CompressedSize) / dfh->UncompressedSize, dfh->UncompressedSize, dfh->CompressedSize);
			}
		}

		// Do the same for the second file, without compression
		{
			std::ifstream file(fileName2, std::ios::binary | std::ios::ate);
			std::streamsize size = file.tellg();
			file.seekg(0, std::ios::beg);
			std::vector<uint8_t> buffer((uint32_t)size);
			file.read((char *)(buffer.data()), size);

			df.Add("File2", buffer, false);
			FuryUtils::Archive::DatHeader *dfh = df.Header(df.EntryCount() - 1);
			if (dfh->IsNotCompressed) {
				printf("%12s\tUncompressed\t%d\n", dfh->FileName, dfh->UncompressedSize);
			}
			else {
				printf("%12s\tCompressed - %d%%\t%d\t%d\n", dfh->FileName, (100 * dfh->CompressedSize) / dfh->UncompressedSize, dfh->UncompressedSize, dfh->CompressedSize);
			}
		}

		// Get a buffer for the completed archive and save it to a file
		std::ofstream outFile(archiveFileName, std::ios::binary | std::ios::trunc);
		std::vector<uint8_t> buffer;
		df.Buffer(buffer);
		outFile.write((char *)(buffer.data()), buffer.size());
		
		return ;
	}
	catch (FuryUtils::Exceptions::Exception e)
	{
		printf("Error:\n\n%d %s\n", e._errorCode, e._errorString.c_str());
		return;
	}
	return;
}

